File: 516-longest-palindromic-subsequence/longest-palindromic-subsequence.cpp
class Solution {
public:
    int solve(string &s, int i, int j, vector<vector<int>> &dp){
        if (i == j) return 1;
        if (i > j) return 0;

        if (dp[i][j] != -1) return dp[i][j];

        if (s[i] == s[j]){
           return dp[i][j] = 2 + solve(s, i+1, j-1, dp);
        }

        int left = solve(s, i+1, j, dp);
        int right = solve(s, i, j-1, dp);

        return dp[i][j] = max(left, right);
    }
    int longestPalindromeSubseq(string s) {
        int n = s.length();
        vector<vector<int>> dp(n, vector<int> (n, -1));
        return solve(s, 0, n-1, dp);
    }
};

File: 516-longest-palindromic-subsequence/longest-palindromic-subsequence.go
func longestPalindromeSubseq(s string) int {
    n := len(s)
    dp := make([][]int, n+2)
    for i := range dp {
        dp[i] = make([]int, n+2)
    }

    for i := n-1; i >= 0; i-- {
        for j := 0; j < n; j++ {
            if s[i] == s[j] {
                dp[i+1][j+1] = 1 + dp[i+2][j]
            } else {
                dp[i+1][j+1] = max(dp[i+2][j+1], dp[i+1][j])
            }
        }
    }

    return dp[1][n]
}